packages/@aws-cdk/aws-stepfunctions-tasks/lib/lambda/invoke.ts

```diff
@@ -161,6 +162,35 @@ export class LambdaInvoke extends sfn.TaskStateBase {
       };
     }
   }
+
+  /**
+   * Determine the ARN(s) to put into the resource field of the generated
+   * IAM policy based on the type of the provided lambda function.
+   *
+   * When invoking Lambda Versions, we need to grant permissions to all
+   * qualifiers. Otherwise in-flight StepFunction executions will fail with
+   * due to missing permissions. This is because a change of the referenced
+   * Version will cause the Policy to remove permissions for the previous
+   * Version - which is currently in-flight.
+   *
+   * @see https://github.com/aws/aws-cdk/issues/17515
+   */
+  private determineResourceArnsForGrantInvoke(lambdaFunction: IFunction): string[] {
+    if (isVersion(lambdaFunction)) {
+      return lambdaFunction.lambda.resourceArnsForGrantInvoke;
```

Comment:
Is `version.resourceArnsForGrantInvoke` *ever* correct, in the face of changing versions?

Isn't it easier to just update the behavior of `Version`?

Reply:
Semantically, I'd expect `lambdaVersion.grantInvoke(role)` to grant permissions for the specified version and not everything. Otherwise I would do `lambdaVersion.grantInvoke(role)`. That's a far reaching change to make.

The idiomatic way to solve this problem in AWS is to use an Alias. I'm on the fence, but because of this I'm more inclined to argue this "works as intended", than to change the behavior of `Version`.

--- 

Maybe we're going about this the wrong way and we need to solve the undeniably bad customer experience differently.

Reply:
@rix0rrr @madeline-k #20177 seems to be of a similar nature. Specifically @kaizencc made the point that if users want to "[..] grant less permissible policies, grant them on the version or alias you've created, not the function itself.".

You've all clearly thought about this in-depth, so I'm very happy to follow your lead.

Reply:
I think Kaizen is right. Users should have created an alias, and invoked that. Invoking by `currentVersion` seems always incorrect, given the code that CDK generates.

In fact we should probably discourage it in the docs?

Reply:
@rix0rrr So close this PR, update the docs, and then close the related issue with explanation?

Reply:
@rix0rrr Can you quickly confirm if I understood you correctly regarding this PR and issue?

> So close this PR, update the docs, and then close the related issue with explanation?

Reply:
Agreed with @rix0rrr that this PR does not address the root cause of the issue.